src/ORM/LazyEagerLoader.php

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Contains methods that are capable of injecting eagerly loaded associations into
+ * entities or lists of entities by using the same syntax as the EagerLoader.
+ *
+ * @internal
+ */
+class LazyEagerLoader
+{
+
+    /**
+     * Loads the specified associations in the passed entity or list of entities
+     * by executing extra queries in the database and merging the results in the
+     * appropriate properties.
+     *
+     * The properties for the associations to be loaded will be overwritten on each entity.
+     *
+     * @param Cake\Datasource\EntityInterface|array $entities a single entity or list of entities
+     * @param array $contain A `contain()` compatible array.
+     * @see Cake\ORM\Query\contain()
+     * @param Cake\ORM\Table $source The table to use for fetching the top level entities
+     * @return Cake\Datasource\EntityInterface|array
+     */
+    public function loadInto($entities, array $contain, Table $source)
+    {
+        $returnSingle = false;
+
+        if ($entities instanceof EntityInterface) {
+            $entities = [$entities];
+            $returnSingle = true;
+        }
+
+        $entities = new Collection($entities);
+        $query = $this->_getQuery($entities, $contain, $source);
+        $associations = array_keys($query->contain());
+
+        $entities = $this->_injectResults($entities, $query, $associations, $source);
+        return $returnSingle ? array_shift($entities) : $entities;
+    }
+
+    /**
+     * Builds a query for loading the passed list of entity objects along with the
+     * associations specified in $contain.
+     *
+     * @param Cake\Collection\CollectionInterface $objects
+     * @param array $contain The associations to be loaded
+     * @param Cake\ORM\Table $source The table to use for fetching the top level entities
+     * @return Cake\ORM\Query
+     */
+    protected function _getQuery($objects, $contain, $source)
+    {
+        $primaryKey = $source->primaryKey();
+        $method = is_string($primaryKey) ? 'get' : 'extract';
+
+        $keys = $objects->map(function ($entity) use ($primaryKey, $method) {
+            return $entity->{$method}($primaryKey);
+        });
+
+        $query = $source
+            ->find()
+            ->select((array)$primaryKey)
+            ->where(function ($exp, $q) use ($primaryKey, $keys, $source) {
+                if (is_array($primaryKey) && count($primaryKey) === 1) {
+                    $primaryKey = current($primaryKey);
+                }
+
+                if (is_string($primaryKey)) {
+                    return $exp->in($source->aliasField($primaryKey), $keys->toList());
+                }
+
+                $types = array_intersect_key($q->defaultTypes(), array_flip($primaryKey));
+                $primaryKey = array_map([$source, 'aliasField'], $primaryKey);
+                return new TupleComparison($primaryKey, $keys->toList(), $types, 'IN');
+            })
+            ->contain($contain);
```

Comment:
Yeah, that is one of the concerns. Mostly that some associations (especially `BelongsTo`) can be defined with conditions dependent on the calling table. I wanted to avoid any invalid sql problems at the cost of an extra query